### PR TITLE
[DB-Migration][First Revision]: Added the required new database tables by creating their db models and alembic script.

### PR DESCRIPTION
**Pull request by NehaKembalkarA10**
## Description
Added new database tables required for a10-octavia in a custom database named `a10_octavia` for now, by following the schema diagram(**update_db**) attached in https://a10networks.atlassian.net/browse/STACK-1177, by using the db models and their respective alembic revision script.
Creating the following tables by this:

- amphora_meta
- thunder
- role
- thunder_cluster
- partitions
- project
- ethernet_interface
- trunk_interface
- ve_interface_cluster

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-1723
https://a10networks.atlassian.net/browse/STACK-1725
https://a10networks.atlassian.net/browse/STACK-1740
https://a10networks.atlassian.net/browse/STACK-1741
https://a10networks.atlassian.net/browse/STACK-1742
https://a10networks.atlassian.net/browse/STACK-1758
https://a10networks.atlassian.net/browse/STACK-1759

## Config
Added changes from : https://github.com/a10networks/a10-octavia/pull/222

## Technical Approach
- Created database models for all the newly required db tables in **models.py**
- Modified the existing VThunder and VRID models as per latest db tables requirements to make it in sync with the revisions.
- Added `target_metadata` to **env.py** for autogenerating the alembic revision script as per modified models.py.
- Autogenerated the alembic revision script **b63ad99c9123_create_required_tables.py** for creating all the newly required tables in database.

## Manual Testing
For now, Tested the table creation on a custom database named a10_octavia.
Step1: Create database a10_octavia:
mysql> **CREATE DATABASE a10_octavia;**

Step2: For connecting to the custom db, add following to [database] section in `/etc/octavia/octavia.conf` file and then restart the octavia-api service:
`connection = mysql+pymysql://root:password@127.0.0.1:3306/a10_octavia`

Step2: From the /path/to/a10-octavia/a10_octavia/db/migration folder run:
$ alembic upgrade head

Step3: Check the newly created tables with their added columns in mysql a10_octavia database
mysql> use a10_octavia;

mysql> show tables;
+-----------------------+
| Tables_in_a10_octavia |
+-----------------------+
| alembic_version |
| amphora_meta |
| ethernet_interface |
| partitions |
| project |
| role |
| thunder |
| thunder_cluster |
| trunk_interface |
| ve_interface_cluster |
| vrid |
| vthunders |
+-----------------------+

### CODE REVIEW

File: a10_octavia/db/models.py [hunk 1]
```diff
@@ -60,7 +60,93 @@ class VRID(base_models.BASE):
...
+    __tablename__ = 'amphora_meta'
+
+    id = sa.Column(sa.String(36), primary_key=True, nullable=False)
+    created_at = sa.Column(u'created_at', sa.DateTime(), nullable=True)
+    updated_at = sa.Column(u'updated_at', sa.DateTime(), nullable=True)
+    last_udp_update = sa.Column(u'last_udp_update', sa.DateTime(), nullable=False)
+    status = sa.Column('status', sa.String(36), default='ACTIVE', nullable=False)
+
+
+class Thunder(base_models.BASE):
```
Comment (rahulkatre-a10) [inline a10_octavia/db/models.py:79]: The relation between thunder and thunder_cluster is not established.

Comment (NehaKembalkarA10) [reply a10_octavia/db/models.py:79]: Adding thunder_cluster_id as a foreign key to thunder table to establish the relation.

File: a10_octavia/db/models.py [hunk 2]
```diff
@@ -60,7 +60,93 @@ class VRID(base_models.BASE):
 
     id = sa.Column(sa.Integer, primary_key=True)
     project_id = sa.Column(sa.String(36), nullable=False)
-    vrid = sa.Column(sa.Integer, default=0)
+    vrid = sa.Column(sa.Integer, default=0, nullable=False)
```
[the remark this answers is not in the document]
Comment (NehaKembalkarA10) [reply a10_octavia/db/models.py:63]: There is a revision `4028e5f7a198_added_vrid_table.py` where this change is occurred without making those respective changes to its db model "VRID". So here, when I am autogenerating the revision for `create_required_tables` from the updated models.py, then at that time as the VRID model is not having this **nullable=False** field in it, it is causing alter column operation to occur in the `create_required_tables` revision for the vrid table.
Thus I am making those modifications as per existing revisions to the model also for not getting this issue.

File: a10_octavia/db/models.py [hunk 3]
```diff
@@ -60,7 +60,93 @@ class VRID(base_models.BASE):
...
-    vrid_floating_ip = sa.Column(sa.String(40))
+    vrid_floating_ip = sa.Column(sa.String(40), nullable=False)
     subnet_id = sa.Column(sa.String(36), nullable=False)
+
+
+class Amphora_Meta(base_models.BASE):
+    __tablename__ = 'amphora_meta'
+
+    id = sa.Column(sa.String(36), primary_key=True, nullable=False)
+    created_at = sa.Column(u'created_at', sa.DateTime(), nullable=True)
```
[the remark this answers is not in the document]
Comment (NehaKembalkarA10) [reply a10_octavia/db/models.py]: Done

File: a10_octavia/db/migration/alembic_migrations/versions/b63ad99c9123_create_required_tables.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                    sa.Column('undercloud', sa.Boolean(), nullable=False),
+                    sa.PrimaryKeyConstraint('id')
+                    )
+    op.create_table('trunk_interface',
+                    sa.Column('interface_num', sa.Integer(), nullable=False),
+                    sa.Column('vlan_id', sa.String(length=36), nullable=False),
+                    sa.Column('subnet_id', sa.String(length=36), nullable=False),
+                    sa.Column('ve_ip_address', sa.String(length=64), nullable=False),
+                    sa.Column('port_id', sa.String(length=36), nullable=False),
+                    sa.Column('state', sa.String(length=36), nullable=False),
```
Comment (afrin-chakure-a10) [inline a10_octavia/db/migration/alembic_migrations/versions/b63ad99c9123_create_required_tables.py:61]: As discussed during call, remove state.

File: a10_octavia/db/migration/alembic_migrations/versions/b63ad99c9123_create_required_tables.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                    sa.PrimaryKeyConstraint('id')
+                    )
+    op.create_table('role',
+                    sa.Column('id', sa.String(length=36), nullable=False),
+                    sa.Column('role', sa.String(length=36), nullable=False),
+                    sa.Column('thunder_id', sa.String(length=36), nullable=True),
+                    sa.ForeignKeyConstraint(['thunder_id'], ['thunder.id'], ),
+                    sa.PrimaryKeyConstraint('id')
+                    )
+    op.create_table('ve_interface_cluster',
```
Comment (afrin-chakure-a10) [inline a10_octavia/db/migration/alembic_migrations/versions/b63ad99c9123_create_required_tables.py:91]: Rename this to `device_network_cluster`

File: a10_octavia/db/migration/alembic_migrations/versions/b63ad99c9123_create_required_tables.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                    sa.Column('id', sa.String(length=36), nullable=False),
+                    sa.Column('vcs_device_id', sa.String(length=1), nullable=False),
+                    sa.Column('management_ip_address', sa.String(length=64), nullable=False),
+                    sa.Column('cluster_id', sa.String(length=36), nullable=True),
+                    sa.ForeignKeyConstraint(['cluster_id'], ['thunder_cluster.id'], ),
+                    sa.PrimaryKeyConstraint('id')
+                    )
+    op.create_table('role',
+                    sa.Column('id', sa.String(length=36), nullable=False),
+                    sa.Column('role', sa.String(length=36), nullable=False),
```
Comment (afrin-chakure-a10) [inline a10_octavia/db/migration/alembic_migrations/versions/b63ad99c9123_create_required_tables.py:86]: change this to `state`

File: a10_octavia/db/models.py [hunk 7]
```diff
@@ -60,7 +66,92 @@ class VRID(base_models.BASE):
...
+
+class Trunk_Interface(base_models.BASE):
+    __tablename__ = 'trunk_interface'
+
+    interface_num = sa.Column(sa.Integer, primary_key=True, nullable=False)
+    vlan_id = sa.Column(sa.String(36), nullable=False)
+    subnet_id = sa.Column(sa.String(36), nullable=False)
+    ve_ip_address = sa.Column(sa.String(64), nullable=False)
+    port_id = sa.Column(sa.String(36), nullable=False)
+    state = sa.Column(sa.String(36), nullable=False)
```
Comment (afrin-chakure-a10) [inline a10_octavia/db/models.py:147]: remove state

File: a10_octavia/db/models.py [hunk 8]
```diff
@@ -60,7 +66,92 @@ class VRID(base_models.BASE):
...
+    cluster_ip_address = sa.Column(sa.String(64), nullable=False)
+    topology = sa.Column(sa.String(50))
+    undercloud = sa.Column(sa.Boolean(), default=False, nullable=False)
+
+
+class Partitions(base_models.BASE):
+    __tablename__ = 'partitions'
+
+    id = sa.Column(sa.String(36), primary_key=True, nullable=False)
+    partition_name = sa.Column(sa.String(14), nullable=True)
```
Comment (afrin-chakure-a10) [inline a10_octavia/db/models.py:116]: change this to `name`

**Review by afrin-chakure-a10 — COMMENTED** [review]
From latest discussion with Team

File: a10_octavia/db/migration/alembic_migrations/versions/b63ad99c9123_create_required_tables.py (hunk 4 above)
Comment (NehaKembalkarA10) [reply a10_octavia/db/migration/alembic_migrations/versions/b63ad99c9123_create_required_tables.py:61]: Done

File: a10_octavia/db/migration/alembic_migrations/versions/b63ad99c9123_create_required_tables.py (hunk 5 above)
Comment (NehaKembalkarA10) [reply a10_octavia/db/migration/alembic_migrations/versions/b63ad99c9123_create_required_tables.py:91]: done

File: a10_octavia/db/migration/alembic_migrations/versions/b63ad99c9123_create_required_tables.py (hunk 6 above)
Comment (NehaKembalkarA10) [reply a10_octavia/db/migration/alembic_migrations/versions/b63ad99c9123_create_required_tables.py:86]: done

File: a10_octavia/db/models.py (hunk 7 above)
Comment (NehaKembalkarA10) [reply a10_octavia/db/models.py:147]: done

File: a10_octavia/db/models.py (hunk 8 above)
Comment (NehaKembalkarA10) [reply a10_octavia/db/models.py:116]: done

**Review by afrin-chakure-a10 — APPROVED** [review]

**Pull request merged**